odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2372 @@
+
+            This option cannot be combined with ``dist``,
+            ``norm`` or ``inner``.
+
+            Default: no weighting
+
+        exponent : positive float, optional
+            Exponent of the norm. For values other than 2.0, no
+            inner product is defined.
+
+            This option is ignored if ``dist``, ``norm`` or
+            ``inner`` is given.
+
+            Default: 2.0
+
+        Other Parameters
+        ----------------
+        dist : callable, optional
+            The distance function defining a metric on the space.
+            It must accept two `CupyTensor` arguments and
+            fulfill the following mathematical conditions for any
+            three vectors ``x, y, z``:
+
+            - ``dist(x, y) >= 0``
+            - ``dist(x, y) = 0``  if and only if  ``x = y``
+            - ``dist(x, y) = dist(y, x)``
+            - ``dist(x, y) <= dist(x, z) + dist(z, y)``
+
+            This option cannot be combined with ``weight``,
+            ``norm`` or ``inner``.
+
+        norm : callable, optional
+            The norm implementation. It must accept an
+            `CupyTensor` argument, return a float and satisfy the
+            following conditions for all vectors ``x, y`` and scalars
+            ``s``:
+
+            - ``||x|| >= 0``
+            - ``||x|| = 0``  if and only if  ``x = 0``
+            - ``||s * x|| = |s| * ||x||``
+            - ``||x + y|| <= ||x|| + ||y||``
+
+            By default, ``norm(x)`` is calculated as ``inner(x, x)``.
+
+            This option cannot be combined with ``weight``,
+            ``dist`` or ``inner``.
+
+        inner : callable, optional
+            The inner product implementation. It must accept two
+            `CupyTensor` arguments, return a element from
+            the field of the space (real or complex number) and
+            satisfy the following conditions for all vectors
+            ``x, y, z`` and scalars ``s``:
+
+            - ``<x, y> = conj(<y, x>)``
+            - ``<s*x + y, z> = s * <x, z> + <y, z>``
+            - ``<x, x> = 0``  if and only if  ``x = 0``
+
+            This option cannot be combined with ``weight``,
+            ``dist`` or ``norm``.
+
+        kwargs :
+            Further keyword arguments are passed to the weighting
+            classes.
+
+        Examples
+        --------
+        Initialization with the class constructor:
+
+        >>> space = CupyTensorSpace(3, 'float')
+        >>> space
+        rn(3, impl='cupy')
+        >>> space.shape
+        (3,)
+        >>> space.dtype
+        dtype('float64')
+
+        A more convenient way is to use the factory functions with the
+        ``impl='cupy'`` option:
+
+        >>> space = odl.rn(3, impl='cupy', weighting=[1, 2, 3])
+        >>> space
+        rn(3, impl='cupy', weighting=[1, 2, 3])
+        >>> space = odl.tensor_space((2, 3), impl='cupy', dtype=int)
+        >>> space
+        tensor_space((2, 3), 'int', impl='cupy')
```

Comment:
Won't these doctests fail misserably for users without cupy, or do we exclude them somehow?

Reply:
At least not when running the file. I'll have to check how to configure `pytest` to exclude them as well.

Reply:
Well if you explicitly run this file, you should get errors IMO, its harder with the global tests

Reply:
Done